### PR TITLE
Bug 1352470: actually delete stuff

### PR DESCRIPTION
**Pull request by djmitche**
Aside from the review, can you double-check my logic that this won't, uh, "accidentally" delete all the exchanges and queues on the pulse server?

The production pulse user has its "config" set appropriately to prohibit deleting non-managed exchanges and queues, but as an administrator using the management API I don't know if that permission applies.

I could land a variant of this which just logs the queues and exchanges it wants to delete, and then try it out.  But I think the unit test already does that.  ??

### CODE REVIEW

File: src/rabbitmanager.js [hunk 1]
```diff
@@ -128,6 +136,51 @@ class RabbitManager {
...
+  }
+
+  /**
+   * Delete an exchange
+   *
+   * @param {string} name       - Name of the exchange.
+   * @param {string} vhost      - Virtual host.
+   * @default
+   */
+  async deleteExchange(name, vhost='/') {
```
Comment (imbstack) [inline src/rabbitmanager.js]: Forgive my ignorance of rabbitmq. Do we really end up with a path of `exchanges///${name}` by default?

File: src/maintenance.js [hunk 2]
```diff
@@ -0,0 +1,159 @@
...
+  }
+
+  return newNamespace;
+};
+
+module.exports.delete = async function({Namespace, rabbitManager, cfg, namespace}) {
+  let debug = Debug('delete');
+
+  // use the configuration regexp to determine if an object is owned by this user
+  let owned = new RegExp(cfg.app.userConfigPermission.replace(/{{namespace}}/g, namespace));
```
Comment (imbstack) [inline src/maintenance.js]: How much guarding should we be doing on this namespace field? If it is accidentally an empty string or something will bad stuff happen?

File: src/maintenance.js [hunk 3]
```diff
@@ -0,0 +1,159 @@
...
+  let debug = Debug('rotate');
+
+  await Namespace.scan({
+    nextRotation: Namespace.op.lessThan(now),
+  }, {
+    limit:            250, // max number of concurrent modify operations
+    handler:          async (ns) => {
+      count++;
+      debug(`rotating ${ns.namespace}`);
+      let password = slugid.v4();
```
Comment (imbstack) [inline src/maintenance.js]: Now that I think about it, we don't make any crypto guarantees about slugids in the slugid readme, afaict. Is it a safe generator to use for this purpose? I'm 99.9% sure it is fine given the use case, but perhaps we should make a note of our reasoning somewhere.

File: src/rabbitmanager.js (hunk 1 above)
Comment (djmitche) [inline src/rabbitmanager.js]: Better, `exchanges/%2F/${name}`

File: src/maintenance.js (hunk 2 above)
Comment (djmitche) [inline src/maintenance.js]: Good thinking.  We guard it at user input - https://github.com/taskcluster/taskcluster-pulse/blob/master/src/v1.js#L193

It wouldn't hurt to re-verify at least the character set here, though.

File: src/maintenance.js (hunk 3 above)
Comment (djmitche) [inline src/maintenance.js]: I think @petemoore has thought quite a bit about that, at least about the entropy.  I recall verifying at one point that we are using a CSPRNG.  Maybe we should just update the slugid docs.  I'll put that on my list.

**Comment by djmitche** [comment]
> Normal RabbitMQ permissions still apply to monitors and administrators; just because a user is a monitor or administrator does not give them full access to exchanges, queues and bindings through either AMQP or the management plugin. 
-- https://www.rabbitmq.com/management.html

I'll create an expired namespace and run the expire operation manually, but given the above I'm not too terrified.  I'll add a comment about this, too.

**Comment by djmitche** [comment]
I landed https://github.com/taskcluster/taskcluster-pulse/commit/993ed1b6f673f333e6165cabe6cf7854d5a482d2 as a bustage fix

File: src/maintenance.js (hunk 3 above)
Comment (petemoore) [inline src/maintenance.js]: In the [go client](https://github.com/taskcluster/slugid-go/#rng-characteristics), we do, but I don't think we documented it in the node client.

**Pull request merged**